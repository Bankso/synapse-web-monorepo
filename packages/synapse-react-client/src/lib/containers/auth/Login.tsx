import * as React from 'react'
import { Form } from 'react-bootstrap'
import ButtonWithIcon from '../../assets/ButtonWithIcon'
import { SynapseClient } from '../../utils'
import {
  getEndpoint,
  BackendDestinationEnum,
} from '../../utils/functions/getEndpoint'
import { GoogleIcon24 } from '../../assets/GoogleIcon24'
import { Button, IconButton, Link } from '@mui/material'
import IconSvg from '../IconSvg'
import FullWidthAlert from '../FullWidthAlert'
import { useState } from 'react'
import { SynapseClientError } from '../../utils/SynapseClientError'
import useLogin from '../../utils/hooks/useLogin'
import { TwoFactorAuthErrorResponse } from '../../utils/synapseTypes/ErrorResponse'
import TOTPForm from './TOTPForm'

export const PROVIDERS = {
  GOOGLE: 'GOOGLE_OAUTH_2_0',
  ORCID: 'ORCID',
}

type Props = {
  ssoRedirectUrl?: string
  sessionCallback: () => void // Callback is invoked after login
  registerAccountUrl?: string
  resetPasswordUrl?: string
  /* Invoked before redirecting to Google. Useful in portals where we may want to store the current URL to redirect back here. */
  onBeginOAuthSignIn?: () => void
  handleIsOnUsernameOrPasswordScreen?: React.Dispatch<
    React.SetStateAction<boolean>
  >
  showUsernameOrPassword?: boolean | undefined
  renderBackButton?: boolean
  /* optionally pass the 2FA error response to directly start the 2FA challenge */
  twoFactorAuthenticationRequired?: TwoFactorAuthErrorResponse
}

/**
 *  Demo of user session, show login screen and handling user login submission.
 *
 *  To support Google SSO in your portal, you must add your domain to the Authorized Redirect URIs
 *  for Synapse authentication.
 *  This can be done by visiting https://sagebionetworks.jira.com/servicedesk/customer/portal/9 to set up a collaboration.
 *  Synapse engineers must add your redirect URL in the Google API console found at https://console.cloud.google.com/ for this functionality to work.
 */
function Login(props: Props) {
  const [username, setUsername] = useState('')
  const [password, setPassword] = useState('')

  const {
    ssoRedirectUrl,
    sessionCallback,
    registerAccountUrl = `${getEndpoint(
      BackendDestinationEnum.PORTAL_ENDPOINT,
    )}#!RegisterAccount:0`,
    resetPasswordUrl = `${getEndpoint(
      BackendDestinationEnum.PORTAL_ENDPOINT,
    )}#!PasswordReset:0`,
  } = props

  const {
    step,
    setStep,
    submitUsernameAndPassword,
    submitOneTimePassword,
    errorMessage,
  } = useLogin(sessionCallback, props.twoFactorAuthenticationRequired)

  const renderBackButton =
    props.renderBackButton &&
    (step === 'USERNAME_PASSWORD' || step === 'VERIFICATION_CODE')

  /**
   * Handle user login on click
   */
  async function handleLogin(clickEvent: React.FormEvent<HTMLElement>) {
    clickEvent.preventDefault() // avoid page refresh
    await submitUsernameAndPassword(username, password)
  }

  function onGoogleSignIn(event: React.MouseEvent<HTMLButtonElement>) {
    if (props.onBeginOAuthSignIn) {
      props.onBeginOAuthSignIn()
    }

    event.preventDefault()
    const redirectUrl = ssoRedirectUrl
      ? `${ssoRedirectUrl}${PROVIDERS.GOOGLE}`
      : `${SynapseClient.getRootURL()}?provider=${PROVIDERS.GOOGLE}`
    SynapseClient.oAuthUrlRequest(PROVIDERS.GOOGLE, redirectUrl)
      .then(data => {
        // Send the user to the authorization URL
        window.location = data.authorizationUrl as unknown as Location
      })
      .catch((err: SynapseClientError) => {
        console.log('Error on oAuth url ', err)
      })
  }

  return (
    <div
      id="loginPage"
      className="container LoginComponent SRC-syn-border-spacing bootstrap-4-backport"
    >
      {renderBackButton && (
        <IconButton type="button" onClick={() => setStep('CHOOSE_AUTH_METHOD')}>
          <IconSvg icon="arrowBack" />
        </IconButton>
      )}
      <div className={step === 'CHOOSE_AUTH_METHOD' ? '' : 'hide-component'}>
        <form>
          <ButtonWithIcon
            variant="white"
            onClick={onGoogleSignIn}
            className={`SRC-signin-button`}
            icon={<GoogleIcon24 />}
          >
            Sign in with Google
          </ButtonWithIcon>
        </form>
        <ButtonWithIcon
          variant="white"
          className={`SRC-signin-button`}
          icon={<IconSvg icon="email" />}
          onClick={() => setStep('USERNAME_PASSWORD')}
        >
          Sign in with your email
        </ButtonWithIcon>
      </div>
      <Form
        className={step === 'USERNAME_PASSWORD' ? '' : 'hide-component'}
        onSubmit={e => {
          handleLogin(e)
        }}
      >
        <label htmlFor={'username'}>Username or Email Address</label>
        <Form.Control
          required
          autoComplete="username"
          placeholder="Username or Email Address"
          className="LoginComponent__Input"
          id="username"
          name="username"
          type="text"
          value={username}
          onChange={e => setUsername(e.target.value)}
        />
        <label htmlFor={'current-password'}>Password</label>
        <Form.Control
          required
          autoComplete="current-password"
          placeholder="Password"
          className="LoginComponent__Input"
          id="current-password"
          name="password"
          type="password"
          value={password}
          onChange={e => setPassword(e.target.value)}
        />
        <Link href={resetPasswordUrl}>Forgot password?</Link>
        <Button
          onClick={e => {
            handleLogin(e)
          }}
          type="submit"
          color="primary"
          variant="contained"
          className="SRC-login-button SRC-marginBottomTen"
        >
          Sign in
        </Button>
      </Form>
      {step === 'VERIFICATION_CODE' && (
        <TOTPForm
          onSubmit={totp => {
            submitOneTimePassword(totp)
          }}
        />
      )}
      {(step === 'CHOOSE_AUTH_METHOD' || step === 'USERNAME_PASSWORD') && (
        <div className={'SRC-center-text'}>
          <Link href={registerAccountUrl}>
            Don&apos;t have an account? Create one now
          </Link>
        </div>
      )}
      {errorMessage && (
        <FullWidthAlert
          variant={'warning'}
          isGlobal={false}
          description={errorMessage}
        />
      )}
    </div>
  )
}
export default Login
