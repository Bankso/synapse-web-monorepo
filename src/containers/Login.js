import React from "react";

/*
    Basic login form to demonstrate Synapse login API call and session token retrieval.
*/
class Login extends React.Component {
    
    constructor (props) {
        super(props)
        this.state =  {
            username: '',
            password: '',
            isSignedIn: false,
            token: '',
            hasLoginInFailed: false
        }
        this.handleChange = this.handleChange.bind(this);
        this.handleLogin = this.handleLogin.bind(this);
        this.showToken = this.showToken.bind(this)
        this.showLoginFailure = this.showLoginFailure.bind(this)
    }

    handleChange(event) {
        const target = event.target
        const name = target.name
        const value = target.value
        this.setState(
            { [name]: value}
        );
    }

    handleLogin(event) {
        event.preventDefault()
        this.props.loginEndpoint(this.state.username, this.state.password).then(
            data => {
                this.setState({
                    token: data.sessionToken,
                    isSignedIn: true,
                    hasLoginInFailed: false
                })
            }
        ).catch(
            err => {
                console.log('Failed to login with error\n', err)
                this.setState({
                    hasLoginInFailed: true
                })
            }
        );
    }

    showToken() {
        if (this.state.isSignedIn && this.state.token !== '' && !this.state.hasLoginInFailed) {
            return (<p> Your session token is {this.state.token} </p>)
        }
    }

    showLoginFailure() {
        if (this.state.hasLoginInFailed) {
            return (
                <div>
                    <small className="form-text text-danger">Error: Not a invalid username/password </small>
                    <div className="invalid-feedback" />
                </div>
            )
        }
    }

    showSignInState() {
        if (this.state.isSignedIn) {
            return (
                <p> You are currently <i> not </i> signed in to Synpase </p>
            )
        } else {
            return (
                <p> You are currently <i> signed in. </i> to Synapse </p>
            )
        }
    }

    render () {
        return (
            <div className="container border">
                <p className="text-left"> Sample Login with session token printed to screen </p>
                <p> You are currently <i> {this.state.isSignedIn ? '' : 'not' } </i> {} signed in </p>
                {this.showToken()}
                <form onSubmit={this.handleLogin}>
                    <div className="form-group">
                        <label className="text-left" htmlFor="exampleEmail">
                            Synapse Email/Username: 
                        </label>
                        <input placeholder="Enter email" className="form-control" id="exampleEmail" name="username" type="email" value={this.state.username} onChange={this.handleChange} />
                    </div>
                    <div className="form-group">
                        <label htmlFor="examplePassword">
                            Password: 
                        </label>
                        <input placeholder="Enter password" className="form-control" id="examplePassword" name="password" type="password" value={this.state.password} onChange={this.handleChange} />
                    </div>
                    {this.showLoginFailure()}
                    <button onSubmit={this.handleLogin} type="submit" className="btn btn-primary">Submit</button>
                </form>
            </div>
        )
    }
}

export default Login;