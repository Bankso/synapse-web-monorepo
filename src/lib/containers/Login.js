import React from "react";

/**
 *  Demo of user session, show login screen and handling user login submission.
 *
 * @class Login
 * @extends {React.Component}
 */
class Login extends React.Component {

    /**
     * Creates a user session, maintaining credentials
     * @param {*} props
     * @memberof Login
     */
    constructor(props) {
        super(props)
        this.state = {
            username: '',
            password: '',
            isSignedIn: false,
            hasLoginInFailed: false,
            errorMessage: '',
            dissmissButtonClicked: false
        }
        this.handleChange = this.handleChange.bind(this);
        this.handleLogin = this.handleLogin.bind(this);
        this.getTokenView = this.getTokenView.bind(this)
        this.getLoginFailureView = this.getLoginFailureView.bind(this)
        this.getSignInStateView = this.getSignInStateView.bind(this)
    }

    /**
     * Updates internal state with the event that was triggered
     *
     * @param {*} event Form update
     */
    handleChange(event) {
        const target = event.target
        const name = target.name
        const value = target.value
        this.setState(
            { [name]: value }
        );
    }

    /**
     * Handle user login on click
     *
     * @param {*} clickEvent Userclick event
     */
    handleLogin(clickEvent) {
        clickEvent.preventDefault()  // avoid page refresh
        this.props.loginEndpoint(this.state.username, this.state.password).then(
            data => {
                if (data.reason) {
                    // error in callback
                    throw new Error(data.reason)
                } else {
                    this.props.onTokenChange({ token: data.sessionToken })
                    this.setState({
                        isSignedIn: true,
                        hasLoginInFailed: false
                    })
                }
            }
        ).catch(
            err => {
                this.setState({
                    hasLoginInFailed: true,
                    errorMessage: err.reason
                })
            }
        )
    }

    /**
     * Shows user session token if they've signed in
     *
     * @returns View displaying user session on login, otherwise null.
     */
    getTokenView() {
        if (this.state.isSignedIn && this.props.token !== '' && !this.state.hasLoginInFailed) {
            return (<p> Your session token is {this.props.token} </p>)
        }
    }

    /**
     * Shows user login failure view on login failure
     *
     * @returns view to be displayed on user sign in error.
     */
    getLoginFailureView() {
        if (this.state.hasLoginInFailed) {
            return (
                <div>
                    <small className="form-text text-danger">  {this.state.errorMessage} </small>
                    <div className="invalid-feedback" />
                </div>
            )
        }
    }

    /**
     * Show whether user is signed in or not, display banner on login success
     *
     * @returns View corresponding to whether the user is signed in, whether they've dismissed
     * sign in banner
     */
    getSignInStateView() {
        if (!this.state.isSignedIn) {
            return (
                <p> You are currently <strong> <i> not </i> </strong> signed in to Synpase </p>
            )
        } else if (!this.state.dissmissButtonClicked){
            return (
                <div>
                    <p> You are currently <strong> <i> signed in </i> </strong> to Synapse </p>
                    <div className="alert alert-success alert-dismissible fade show" role="alert">
                        Synapse login successfull
                        <button type="button" className="close" onClick={() => {this.setState({dissmissButtonClicked: true})}}>
                            <span aria-hidden="true">&times;</span>
                        </button>
                    </div>
                </div>
            )
        }
    }

    render() {
        return (
            <div id="loginPage" className="container border">
                <h3 className="text-left"> Demo login with session token printed to screen</h3>
                {this.getSignInStateView()}
                {this.getTokenView()}
                <form autoComplete="on" onSubmit={this.handleLogin}>
                    <div className="form-group">
                        <label className="text-left" htmlFor="exampleEmail">
                            Synapse Email/Username:
                        </label>
                        <input placeholder="Enter email" className="form-control" id="exampleEmail" name="username" type="email" value={this.state.username} onChange={this.handleChange} />
                    </div>
                    <div className="form-group">
                        <label htmlFor="examplePassword">
                            Password:
                        </label>
                        <input placeholder="Enter password" className="form-control" id="examplePassword" name="password" type="password" value={this.state.password} onChange={this.handleChange} />
                    </div>
                    {this.getLoginFailureView()}
                    <button onSubmit={this.handleLogin} type="submit" className="btn btn-primary m-1">Submit</button>
                </form>
            </div>
        )
    }
}

export default Login;