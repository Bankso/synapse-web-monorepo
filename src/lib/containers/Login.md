```jsx
<Login sessionCallback={()=>{location.reload()}} theme={'light'} icon={true} />
```