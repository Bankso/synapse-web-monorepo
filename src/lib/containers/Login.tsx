import * as React from 'react'
import ButtonContent from '../assets/ButtonContent'
import GoogleIcon from '../assets/GoogleIcon'
import { SynapseClient } from '../utils'

type State = {
  username: string
  password: string
  email: string
  isSignedIn: boolean
  hasLoginInFailed: boolean
  errorMessage: string
}

type Props = {
  token: string | undefined
  theme: string
  icon: boolean
}
export const AUTH_PROVIDER = 'GOOGLE_OAUTH_2_0'
export const getRootURL = () => {
  return `${window.location.protocol}//${window.location.hostname}:${window.location.port}/`
}

/**
 *  Demo of user session, show login screen and handling user login submission.
 *
 *  To support Google SSO in your portal, you must add your domain to the Authorized Redirect URIs
 *  for Synapse authentication.
 *  This can be done by contacting synapseInfo@sagebionetworks.org to form a collaboration.
 *  Synapse engineers must add your redirect URL in the Google API console found at https://console.cloud.google.com/ for this functionality to work.
 *
 * @class Login
 * @extends {React.Component}
 */
class Login extends React.Component<Props, State> {

    /**
     * Creates a user session, maintaining credentials
     * @param {*} props
     * @memberof Login
     */
  constructor(props: Props) {
    super(props)
    this.state = {
      email: '',
      errorMessage: '',
      hasLoginInFailed: false,
      isSignedIn: false,
      password: '',
      username: ''
    }
    this.handleChange = this.handleChange.bind(this)
    this.handleLogin = this.handleLogin.bind(this)
    this.getTokenView = this.getTokenView.bind(this)
    this.getLoginFailureView = this.getLoginFailureView.bind(this)
    this.onSignIn = this.onSignIn.bind(this)
  }
    /**
     * Updates internal state with the event that was triggered
     *
     * @param {*} event Form update
     */
  public handleChange(event: React.FormEvent<HTMLInputElement>): void {
    const target = event.currentTarget
    const name = target.name
    const value = target.value
    const newState: (Pick<any, any>) = { [name]: value }
    this.setState(newState)
  }
    /**
     * Handle user login on click
     *
     * @param {*} clickEvent Userclick event
     */
  public handleLogin(clickEvent: React.FormEvent<HTMLElement>) {
    clickEvent.preventDefault() // avoid page refresh
    SynapseClient.login(this.state.username, this.state.password)
            .then((data: any) => {
              SynapseClient.setSessionTokenCookie(data.sessionToken).then(() => {
                // on session change, reload the page so that all components get the new token from the cookie
                window.location.reload()
              }).catch((errSetSession) => {
                console.log('Could not set session token cookie!', errSetSession)
                this.setState({
                  errorMessage: errSetSession.reason,
                  hasLoginInFailed: true,
                  isSignedIn: false
                })
              })
            })
            .catch((err: any) => {
              console.log('Handle login failed with err = ', err)
              this.setState({
                errorMessage: err.reason,
                hasLoginInFailed: true,
                isSignedIn: false
              })
            })
  }

    /**
     * Shows user session token if they've signed in
     *
     * @returns View displaying user session on login, otherwise null.
     */
  public getTokenView(): (JSX.Element | boolean) {
    if (this.state.isSignedIn && this.props.token !== '' && !this.state.hasLoginInFailed) {
      return <p> Your session token is {this.props.token} </p>
    }
    return false
  }
    /**
     * Shows user login failure view on login failure
     *
     * @returns view to be displayed on user sign in error.
     */
  public getLoginFailureView(): (JSX.Element | boolean) {
    if (this.state.hasLoginInFailed) {
      return (
                <div>
                    <small className="form-text text-danger"> {this.state.errorMessage} </small>
                    <div className="invalid-feedback" />
                </div>
      )
    }
    return false
  }
    /**
     * Show whether user is signed in or not, display banner on login success
     *
     * @returns View corresponding to whether the user is signed in, whether they've dismissed
     * sign in banner
     */
  public getSignInStateView(): (JSX.Element | boolean) {
    if (!this.state.isSignedIn) {
      return (
                <p>
                    {' '}
                    You are currently{' '}
                    <strong>
                        {' '}
                        <i> not </i>{' '}
                    </strong>{' '}
                    signed in to Synpase{' '}
                </p>
      )
    }
    return false
  }
  public onSignIn(event: React.MouseEvent<HTMLButtonElement>) {
    // save current route (so that we can go back here after SSO)
    localStorage.setItem('after-sso-login-url', window.location.href)
    event.preventDefault()
    SynapseClient.oAuthUrlRequest(AUTH_PROVIDER, `${getRootURL()}?provider=${AUTH_PROVIDER}`)
            .then((data: any) => {
              const authUrl = data.authorizationUrl
              window.location = authUrl // ping the url
            })
            .catch((err: any) => {
              console.log('Error on oAuth url ', err)
            })
  }
  public render() {
    const { theme, icon } = this.props
    const googleTheme = theme === 'dark' ? 'SRC-google-button-dark-color' : 'SRC-google-button-light-color'
    return (
            <div id="loginPage" className="container loginContainer SRC-syn-border-spacing">
                  <form>
                    {/* tslint:disable-next-line */}
                    <button onClick={this.onSignIn} className={`SRC-google-button ${googleTheme} SRC-marginBottomTen`}>
                        <GoogleIcon key={1} active={true} />
                        <ButtonContent icon={icon} key={2}>
                            Sign in with Google
                        </ButtonContent>
                    </button>
                </form>
                <div className="SRC-center-text SRC-deemphasized-text SRC-marginBottomTen">or</div>
                <div className="SRC-center-text SRC-marginBottomTen">Sign in with your Synapse account</div>
                <form onSubmit={this.handleLogin}>
                    <div className="form-group">
                        <input
                                autoComplete="email"
                                placeholder="username or email"
                                className="form-control"
                                id="exampleEmail"
                                name="username"
                                type="text"
                                value={this.state.username}
                                onChange={this.handleChange}
                                data-lpignore="true"
                        />
                    </div>
                    <div className="form-group">
                        <input
                                autoComplete="password"
                                placeholder="password"
                                className="form-control"
                                id="examplePassword"
                                name="password"
                                type="password"
                                value={this.state.password}
                                onChange={this.handleChange}
                                data-lpignore="true"
                        />
                    </div>
                    {this.getLoginFailureView()}
                    <button
                      onSubmit={this.handleLogin}
                      type="submit"
                      className="btn SRC-primary-background-color SRC-hoverWhiteText
                        SRC-whiteText m-1 SRC-google-button SRC-marginBottomTen"
                    >
                      <ButtonContent>
                        Sign in
                      </ButtonContent>
                    </button>
                </form>
                <div>
                  <a
                    href="https://www.synapse.org/#!PasswordReset:0"
                    className="SRC-floatLeft"
                  >
                    Forgot password?
                  </a>
                  <span className="SRC-deemphasized-text SRC-floatRight">&nbsp;It's free!</span>
                  <a
                    href="https://www.synapse.org/#!RegisterAccount:0"
                    className="SRC-floatRight"
                  >
                    Register
                  </a>
                </div>
            </div>
    )
  }
}
export default Login
