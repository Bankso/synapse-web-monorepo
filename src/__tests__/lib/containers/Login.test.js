import React from 'react';
import {configure, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';

import Login from '../../../lib/containers/Login';

configure({ adapter: new Adapter() });


describe("test ", () => {
  it('renders correctly', () => {
    const tree = mount(<Login />)
  });
})