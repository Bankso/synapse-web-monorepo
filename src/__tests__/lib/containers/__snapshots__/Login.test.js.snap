// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container syn-example"
  id="loginPage"
>
  <h3
    className="text-left"
  >
     Demo login with session token printed to screen
  </h3>
  <p>
     You are currently 
    <strong>
       
      <i>
         not 
      </i>
       
    </strong>
     signed in to Synpase 
  </p>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className="text-left"
        htmlFor="exampleEmail"
      >
        Synapse Email/Username:
      </label>
      <input
        autoComplete="email"
        className="form-control"
        id="exampleEmail"
        name="username"
        onChange={[Function]}
        placeholder="Enter email"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="examplePassword"
      >
        Password:
      </label>
      <input
        autoComplete="password"
        className="form-control"
        id="examplePassword"
        name="password"
        onChange={[Function]}
        placeholder="Enter password"
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary m-1"
      onSubmit={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
  <form>
    <div
      className="form-group"
      style={
        Object {
          "margin": "2px",
        }
      }
    >
      <button
        className={undefined}
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "background": "#d14836",
            "border": "1px solid transparent",
            "borderRadius": 2,
            "color": "#fff",
            "display": "inline-block",
            "fontFamily": "Roboto",
            "fontSize": 16,
            "fontWeight": "bold",
            "opacity": 0.6,
            "paddingBottom": 10,
            "paddingTop": 10,
            "width": 190,
          }
        }
        type="button"
      >
        Login
      </button>
    </div>
    <button
      className="btn btn-default"
      onClick={[Function]}
      type="button"
    >
       Sign Out 
    </button>
  </form>
</div>
`;
